src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1254 @@
+        local_snap->snapshot_namespace);
+    if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_USER) {
+      bool unlink_user_snap = true;
+      for (auto &remote_snap : m_remote_group_snaps) {
+        if (remote_snap.name == local_snap->name) {
+          unlink_user_snap = false;
+          break;
+        }
+      }
+      if (!unlink_user_snap) {
+        locker.unlock();
+        continue;
+      }
+      dout(10) << "unlinking regular group snap in-progress: "
+               << local_snap->name << ", with id: " << local_snap->id << dendl;
+    } else {
+      // remove mirror peer uuid of the remote snap
+
+      auto next_local_snap = std::next(local_snap);
+      // If next local snap is end, or if it is the syncing in-progress snap,
+      // then we still need this group snapshot.
+      if (next_local_snap == m_local_group_snaps.end() ||
+          (next_local_snap->id == remote_group_snap_id &&
+           next_local_snap->state != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE)) {
+        break;
+      }
+      remove_mirror_peer_uuid(local_snap->id);
+    }
+    locker.unlock();
+    // prune all the image snaps of the group snap locally
+    if (prune_all_image_snapshots(&(*local_snap))) {
+      continue;
+    }
+    dout(10) << "all image snaps are pruned, finally unlinking group snap: "
+             << local_snap->id << dendl;
+    locker.lock();
+    r = librbd::cls_client::group_snap_remove(&m_local_io_ctx,
+        librbd::util::group_header_name(m_local_group_id), local_snap->id);
+    if (r < 0) {
+      derr << "failed to remove group snapshot : "
+           << local_snap->id << " : " << cpp_strerror(r) << dendl;
+    }
+    m_local_group_snaps.erase(local_snap);
+    locker.unlock();
+  }
+}
+
+template <typename I>
+void Replayer<I>::create_regular_snapshot(
+    const std::string &remote_group_snap_name,
+    const std::string &remote_group_snap_id,
+    Context *on_finish) {
+  dout(10) << remote_group_snap_id << dendl;
+  librados::ObjectWriteOperation op;
+  cls::rbd::GroupSnapshot group_snap{
+    remote_group_snap_id, // keeping it same as remote group snap id
+    cls::rbd::GroupSnapshotNamespaceUser{},
+      remote_group_snap_name,
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+  m_local_group_snaps.push_back(group_snap);
+
+  librbd::cls_client::group_snap_set(&op, group_snap);
+  auto comp = create_rados_callback(
+      new LambdaContext([this, on_finish](int r) {
+        handle_create_regular_snapshot(r, on_finish);
+      }));
+  int r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void Replayer<I>::handle_create_regular_snapshot(
+    int r, Context *on_finish) {
+  dout(10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    derr << "error creating local non-primary group snapshot: "
+         << cpp_strerror(r) << dendl;
+  }
+  on_finish->complete(0);
```

Comment:
Should this be `on_finish->complete(r)` ? Why ignore error during `group_snap_set`?